x/ecocredit/migrations/v3/core.go

```diff
@@ -315,6 +318,96 @@ func MigrateState(sdkCtx sdk.Context, storeKey storetypes.StoreKey,
 		return err
 	}
 
+	// - update curator address for baskets
+	// we don't have baskets on mainnet
+
```

Comment:
This PR only contains mainnet migration. We cannot add both mainnet & Redwood migrations here.

Reply:
So the idea was that the manual operations specific to each chain (mainnet and redwood) would be run in the upgrade handler and the state migrations would work for both. In the same way that `recoverFunds` is included in the upgrade handler and run separately from the state migrations, the manual operations specific to each chain would also be included in the upgrade handler and no hard-coded values specific to mainnet would be included in the state migrations.

We would then have two upgrade handlers that each perform manual operations following state migrations, one would have the upgrade name `v4.0.0` (mainnet) and one would have the upgrade name `v4.0.0-redwood`. We can hold off on implementing the upgrade handler for redwood in this pull request if you prefer but we should make sure state migrations will run for both mainnet and redwood and the manual operations for mainnet are separate from the state migrations.

Reply:
So the idea was that the manual operations specific to each chain (mainnet and redwood) would be run in the upgrade handler and the state migrations would work for both. In the same way that `recoverFunds` is included in the upgrade handler and run separately from the state migrations, the manual operations specific to each chain would also be included in the upgrade handler and no hard-coded values specific to mainnet would be included in the state migrations.

We would then have two upgrade handlers that each perform manual operations following state migrations, one would have the upgrade name `v4.0.0` (mainnet) and one would have the upgrade name `v4.0.0-redwood`. We can hold off on implementing the upgrade handler for redwood in this pull request but we should make sure state migrations will run for both mainnet and redwood and the manual operations for mainnet are separate from the state migrations.

Reply:
> In the same way that `recoverFunds` is included in the upgrade handler and run separately from the state migrations, the manual operations specific to each chain should be included in upgrade handlers specific to each chain and no hard-coded values specific to one chain and not the other would be included in the state migrations.
> 
I tried a similar approach, I don't think we can access ecocredit state outside. We are using new app wiring for regen modules.

Reply:
Oof. Yea, that sounds about right. I'll try to take a look. cc @clevinson @aaronc

Reply:
We could expose the keeper methods we need and make use of them in the upgrade handler in the same way that we expose `AddCreditType` and `AllowDenom` for custom governance proposals and `PruneOrders` for begin blocker...

Reply:
Not sure that is the right solution though. Let's discuss on our extended standup.

Reply:
We could create and expose a keeper method that would handle the manual operations in the same way that we expose `AddCreditType` and `AllowDenom` for custom governance proposals and `PruneOrders` for begin blocker...

Not sure that is the right solution though. Let's discuss on our extended standup.

Reply:
Ok, I think it makes the most sense to perform the manual operations within the state migrations. We can perform specific operations for each chain by pulling the chain id from sdk context.

It would be nice to separate out the manual operations from the current state migrations and to test them separately. We could call a single function at the end of  `MigrateState` that we define in a separate file that checks the chain id and calls a specific function for each chain id (`regen-1` and `regen-redwood-1`) that performs the manual operations.

Reply:
Ok, I think it makes the most sense to perform the manual operations within the state migrations. We can perform specific operations for each chain by pulling the chain id from sdk context.

It would be nice to separate out the manual operations from the current state migrations and to test them separately. We could call a single function at the end of  `MigrateState` that we define in a separate file that checks the chain id and then calls a specific function for each chain id (`regen-1` and `regen-redwood-1`).

Does that sound like a reasonable solution? I think this is actually an improvement from the original idea @clevinson and I were thinking with separate upgrade handlers. Not sure why we didn't think of it then.

Reply:
Maybe `patch.go` within `migrations/v3` and something like `patchMigration` that would then call either `patchMainnet` or `patchRedwood` based on chain id? If this seems unreasonable, we can further discuss on our extended standup.

Reply:
Maybe `patch.go` within `migrations/v3` and `patchMigration` that would call either `patchMainnet` or `patchRedwood` based on chain id? If this seems unreasonable, we can discuss on our extended standup.

Reply:
Maybe `patch.go` within `migrations/v3` and `patchMigration` that then calls `patchMainnet` or `patchRedwood` based on chain id? If this seems unreasonable, we can further discuss on our extended standup.